apps/internal/oauth/ops/authority/authority.go

```diff
@@ -351,29 +352,31 @@ type Info struct {
 	InstanceDiscoveryDisabled bool
 }
 
-func firstPathSegment(u *url.URL) (string, error) {
-	pathParts := strings.Split(u.EscapedPath(), "/")
-	if len(pathParts) >= 2 {
-		return pathParts[1], nil
-	}
-
-	return "", errors.New(`authority must be an https URL such as "https://login.microsoftonline.com/<your tenant>"`)
-}
-
 // NewInfoFromAuthorityURI creates an AuthorityInfo instance from the authority URL provided.
 func NewInfoFromAuthorityURI(authority string, validateAuthority bool, instanceDiscoveryDisabled bool) (Info, error) {
 	u, err := url.Parse(strings.ToLower(authority))
 	if err != nil || u.Scheme != "https" {
 		return Info{}, errors.New(`authority must be an https URL such as "https://login.microsoftonline.com/<your tenant>"`)
 	}
 
-	tenant, err := firstPathSegment(u)
-	if err != nil {
-		return Info{}, err
+	pathParts := strings.Split(u.EscapedPath(), "/")
+	if len(pathParts) < 2 {
+		return Info{}, errors.New(`authority must be an https URL such as "https://login.microsoftonline.com/<your tenant>"`)
 	}
-	authorityType := AAD
-	if tenant == "adfs" {
+
+	var authorityType, tenant string
+	switch pathParts[1] {
+	case "adfs":
 		authorityType = ADFS
+	case "dstsv2":
+		authorityType = DSTS
+		if len(pathParts) != 3 {
+			return Info{}, errors.New(`dSTS authority must be an https URL such as "https://login.microsoftonline.com/dstsv2/<your tenant>"`)
+		}
+		tenant = pathParts[2]
+	default:
+		authorityType = AAD
+		tenant = pathParts[1]
 	}
 
 	// u.Host includes the port, if any, which is required for private cloud deployments
```

Comment:
I can't comment on the lines below, because I haven't changed them, smh, so I'm going to comment here

### CanonicalAuthorityURI

`u.Host/<tenant>` is not sufficient for dSTS, we should probably use the raw path minus tenant? something like `CanonicalAuthorityURI: strings.Replace(authority, tenant, "")` ... but I am not sure about all the various URLs we can receive, and this might easily lead to malformed URLs

for my use-case, I didn't actually need the `CanonicalAuthorityURI`, so I didn't have to change anything, but we need it to be correct regardless

### UserRealmURIPrefix

as for the `UserRealmURIPrefix`, it is only used in tests, do we want to keep it? if yes, it seems like we should again maybe use the raw path excluding tenant + "/common/userrealm/"